src/template.rs

```diff
@@ -146,81 +121,169 @@ impl Template {
         &self,
         event: impl Into<EventRef<'a>>,
     ) -> Result<String, TemplateRenderingError> {
-        let event = event.into();
-        match (self.has_fields, self.has_ts) {
-            (false, false) => Ok(self.src.clone()),
-            (true, false) => render_fields(&self.src, event),
-            (false, true) => Ok(render_timestamp(&self.src, event)),
-            (true, true) => {
-                let tmp = render_fields(&self.src, event)?;
-                Ok(render_timestamp(&tmp, event))
-            }
+        if self.is_static {
+            Ok(self.src.clone())
+        } else {
+            render_fields(&self.parts, event.into())
         }
     }
 
     pub fn get_fields(&self) -> Option<Vec<String>> {
-        if self.has_fields {
-            RE.captures_iter(&self.src)
-                .map(|c| {
-                    c.get(1)
-                        .map(|s| s.as_str().trim().to_string())
-                        .expect("src should match regex")
-                })
-                .collect::<Vec<_>>()
-                .into()
-        } else {
-            None
-        }
+        let parts: Vec<_> = self
+            .parts
+            .iter()
+            .filter_map(|part| {
+                if let Part::Reference(r) = part {
+                    Some(r.to_owned())
+                } else {
+                    None
+                }
+            })
+            .collect();
+        (!parts.is_empty()).then_some(parts)
+    }
+
+    pub fn get_ref(&self) -> &str {
+        &self.src
+    }
+
+    /// Returns `true` if this template string has a length of zero, and `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.src.is_empty()
     }
 
     pub const fn is_dynamic(&self) -> bool {
-        self.has_fields || self.has_ts
+        !self.is_static
     }
+}
 
-    pub fn get_ref(&self) -> &str {
-        &self.src
+/// One part of the template string after parsing.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+enum Part {
+    /// A literal piece of text to be copied verbatim into the output.
+    Literal(String),
+    /// A literal piece of text containing a time format string.
+    Strftime(ParsedStrftime),
+    /// A reference to the source event, to be copied from the relevant field or tag.
+    Reference(String),
```

Comment:
Once #14887 lands, it would be great to also store the pre-parsed path routing here, but I couldn't see a way to make that work with the existing types.

Reply:
I'll probably just include this in #14887 assuming yours gets merged first.